rfcs/2021-09-21-9292-throttle-transform.md

```diff
@@ -0,0 +1,153 @@
+
+        Box::pin(
+            stream! {
+              loop {
+                let mut output = Vec::new();
+                let done = tokio::select! {
+                    _ = flush_stream.tick() => {
+                        false
+                    }
+                    maybe_event = input_rx.next() => {
+                        match maybe_event {
+                            None => true,
+                            Some(event) => {
+                    if let Some(condition) = self.exclude_as_ref() {
+                    if condition.check(&event) {
+                        output.push(event);
+                    false
+                    }
+                }
+
+                let value = self
+                    .key_field
+                    .as_ref()
+                    .and_then(|key_field| event.get(key_field))
+                    .map(|v| v.to_string_lossy());
+
+                                match lim.check_key_n(value, size) {
+                                    Ok(()) => {
+                                        output.push(event);
+                                        false
+                                    }
+                                    _ => {
+                                        // Dropping event
+                                        false
+                                    }
+                                }
+                            }
+                        }
+                    }
+                };
+                yield stream::iter(output.into_iter());
+                if done { break }
+              }
+            }
+            .flatten(),
+        )
+    }
+}
+```
+
+## Rationale
+
+- Controlling throughput provides better reliability for downstream services which can be critical in observability platforms
+- Setting granular quotas is invaluable as an administrator running Vector for a number of users to ensure fair use of the pipeline
+
+## Drawbacks
+
+- When rate limiting by `Bytes` in a transform it will be the unserialized form of the event, which will differ from what downstream sinks will actually receive
+
+## Prior Art
+
+- [Governor](https://docs.rs/governor/0.3.2/governor/index.html)
+- [Throttle - FluentBit](https://docs.fluentbit.io/manual/pipeline/filters/throttle)
+- [grouper::bucker - Tremor](https://github.com/tremor-rs/tremor-runtime/blob/main/tremor-pipeline/src/op/grouper/bucket.rs)
+
+## Alternatives
+
+- Extend `sample` transform to allow for a window configuration
+- Build rate limiting controls as a generic `sink` feature, rather than a separate `transform`
+
+## Outstanding Questions
+
+- Rate limiting seems like it could be generically a `sink` concern and implemented as a composable part of our `sink` pattern. This could give more "accurate" serialized sizes and possibly be easier to manage for administrators (depending on needs). If rate limiting is also a `sink` concern should it only be implemented there or also available as a `transform`?
```

Comment:
It seems reasonable to implement the transform first. Sinks generally have rate limits available already, just not on a per-key basis and the best implementation there is a little unclear.